tortoise/tortoise.go

```diff
@@ -636,268 +449,264 @@ func (t *turtle) getBallotBeacon(ballot *types.Ballot, logger log.Log) ([]byte,
-		// disagreement about Hare success.
-		if !consensusMatches {
-			logger.With().Warning("global opinion on block differs from our vote, cannot verify layer",
-				bid,
-				log.Stringer("global_opinion", globalOpinionOnBlock),
-				log.Stringer("local_opinion", localOpinionOnBlock))
-			return nil, nil
-		}
+		previous := t.verified
+		t.verified = t.verifying.verifyLayers(logger)
 
-		// There are only two scenarios that could result in a global opinion of abstain: if everyone is still
-		// waiting for Hare to finish for a layer (i.e., it has not yet succeeded or failed), or a balancing attack.
-		// The former is temporary and will go away after `zdist' layers. And it should be true of an entire layer,
-		// not just of a single block. The latter could cause the global opinion of a single block to permanently
-		// be abstain. As long as the contextual validity of any block in a layer is unresolved, we cannot verify
-		// the layer (since the effectiveness of each transaction in the layer depends upon the contents of the
-		// entire layer and transaction ordering). Therefore we have to enter self healing in this case.
-		// TODO: abstain only for entire layer at a time, not for individual blocks (optimization)
-		//   see https://github.com/spacemeshos/go-spacemesh/issues/2674
-		if !globalOpinionDecided {
-			logger.With().Warning("global opinion on block is abstain, cannot verify layer",
-				bid,
-				log.Stringer("global_opinion", globalOpinionOnBlock),
-				log.Stringer("local_opinion", localOpinionOnBlock))
-			return nil, nil
+		if t.verified.After(previous) {
+			if err := persistContextualValidity(logger,
+				t.bdp,
+				previous, t.verified,
+				t.blocks,
+				t.localOpinion,
+			); err != nil {
+				return err
+			}
+		}
+		t.updateHistoricallyVerified()
+		if t.verified == t.processed.Sub(1) {
+			return nil
+		}
+	}
+	// keep votes starting from the layer when verifying tortoise failed to make progress
+	// for example, if verifying failed while counting ballots from layer 13 we will
+	// most likely have to use those ballots later in full mode.
+	//
+	// this may change for rerun https://github.com/spacemeshos/go-spacemesh/issues/2980,
+	// verified layer will be expected to be stuck as the expected weight will be high.
+	// so we will have to use different heuristic for switching into full mode.
+	// maybe the difference between counted and uncounted weight for a specific layer
+	t.keepVotes(ballots)
+	if t.isHealingEnabled() || t.mode == fullMode {
+		if t.mode == verifyingMode {
+			logger.With().Info("switching to full self-healing tortoise",
+				lid,
+				log.Stringer("verified", t.verified),
+			)
+			t.mode = fullMode
+		}
+		previous := t.verified
+		healed := t.healingTortoise(ctx, logger)
+		if !healed {
+			return nil
+		}
+		// TODO(dshulyak) refactor full tortoise to be consistent with verifying.
+		// update local opinion in memory and persist afterwards.
+		t.updateHistoricallyVerified()
```

Comment:
i don't grasp the idea of historically verified. it's updated in both verifying and full mode, how is this different from verified then?

Reply:
it is different during rerun. we are using verified to track progress of the rerun, but local opinion must be based on historically verified.

Reply:
before a rerun starts, `consensus.historicallyVerified` is set to the normal run's `historicallyVerified`. does that mean that rerun will form local opinions based on the normal run's historicallyVerified? i don't understand why rerun cannot form local opinions based on its own value of `verified`

Reply:
so how rerun should work: 
last layer is kept according to the clock
rerun start from genesis and processes all layers up to the last
important: expected weight is set based on the last layer (we don't do it currently https://github.com/spacemeshos/go-spacemesh/issues/2980), because we want to count votes that we didn't count during normal flow. it prevents concentration attacks, or layers that were verified with dishonest majority. it was discussed on last call.
because of the previous heuristic - verified layer will be stuck long in the past. but we also want to use verifying tortoise, this is why we base our local opinion on what we initially verified.

note that the condition i described does happen even now (on data from devnet210, but layer is not stuck far in the past, but may get stuck for 1-2 layers)